Fix phi range

VinInn opened the pull request:
there was a bug in checkPhiInRange and other in PixelTripletHLTGenerator.cc (accepting empty ranges).
Added also a test and two optimized implementations of phi and eta intervals.

regression expected in particular at low-pt
http://innocent.home.cern.ch/innocent/RelVal/pu35_81_fixRP/plots_highPurity/effandfake1.pdf

VinInn committed:
fix detaphi

VinInn committed:
fix phiRange

VinInn committed:
account for range larger than pi

VinInn committed:
change algo, add maxRange

VinInn committed:
skip empty ranges

VinInn committed:
a test

cmsbuild commented:
A new Pull Request was created by  @VinInn (Vincenzo Innocente) for CMSSW_8_1_X.

It involves the following packages:

DataFormats/GeometryVector
DataFormats/Math
RecoPixelVertexing/PixelTriplets


@civanch, @cvuosalo, @mdhildreth, @cmsbuild, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@makortel, @cerati, @GiacomoSguazzoni, @dgulhan, @rovere this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

VinInn commented:
@cmsbuild , please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: a35d7280932b01167ba366f6146f25e38511c204
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13914/12157/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13914/12157/summary.html

civanch commented:
+1

cvuosalo commented on DataFormats/GeometryVector/interface/PhiInterval.h:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  bool inside(float ix, float iy) const {
+    auto norm = 1.f/std::sqrt(ix*ix+iy*iy);
```
Wouldn't it be better to protect against division by zero here and at line 20 in case these methods are called inappropriate zero values?

VinInn replied:
no. too expensive.
better people verify input if in doubt
we should stop to be iper protective and than ask why we spend most of reco time in branches

slava77 replied:
why is 1./sqrt needed here?
isn't it cheaper to ``` return ix*x + iy*y > dcos*sqrt(ix*ix + iy*iy); ```

VinInn replied:
depends...
with Ofast 1.f/std::sqrt is the cheaper than sqrt in float... 
maybe you are right: most probably even with Ofast
return ix*x + iy*y > dcos*sqrt(ix*ix + iy*iy);
has one multiplication less (even without fma)
and parallelize more (less dependencies)

VinInn committed:
Update PhiInterval.h

most probably this is a bit faster even with Ofast or at least not slower (see comment in PR)

cmsbuild commented:
Pull request #13914 was updated. @civanch, @cvuosalo, @mdhildreth, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

VinInn commented on DataFormats/GeometryVector/interface/PhiInterval.h:
```diff
@@ -0,0 +1,42 @@
+    return inside(v.x(),v.y());
+  }
+
+  bool inside(float ix, float iy) const {
```
#include<cmath>
float x,y,dcos;
bool inside(float ix, float iy) {
    return ix*x+iy*y > dcos*std::sqrt(ix*ix+iy*iy);
  }

bool inside2(float ix, float iy) {
  auto norm = 1.f/std::sqrt(ix*ix+iy*iy);
    return norm*(ix*x+iy*y) > dcos;
  }

produces with Ofast
the real point is not the number of mul instruction, but how the out-of-order processor can feed the two FP engines and how the mul are pipelined : I do not think we can easily compute the latency of such a small kernel...

inside(float, float):                            # @inside(float, float)
        movss   x(%rip), %xmm2          # xmm2 = mem[0],zero,zero,zero
        mulss   %xmm0, %xmm2
        movss   y(%rip), %xmm3          # xmm3 = mem[0],zero,zero,zero
        mulss   %xmm1, %xmm3
        addss   %xmm2, %xmm3
        mulss   %xmm0, %xmm0
        mulss   %xmm1, %xmm1
        addss   %xmm0, %xmm1
        xorps   %xmm0, %xmm0
        rsqrtss %xmm1, %xmm0
        movaps  %xmm0, %xmm2
        mulss   %xmm2, %xmm2
        mulss   %xmm1, %xmm2
        addss   .LCPI2_0(%rip), %xmm2
        mulss   .LCPI2_1(%rip), %xmm0
        mulss   %xmm1, %xmm0
        mulss   %xmm2, %xmm0
        xorps   %xmm2, %xmm2
        cmpeqss %xmm2, %xmm1
        andnps  %xmm0, %xmm1
        mulss   dcos(%rip), %xmm1
        ucomiss %xmm1, %xmm3
        seta    %al
        retq

inside2(float, float):                           # @inside2(float, float)
        movaps  %xmm0, %xmm2
        mulss   %xmm2, %xmm2
        movaps  %xmm1, %xmm3
        mulss   %xmm3, %xmm3
        addss   %xmm2, %xmm3
        xorps   %xmm2, %xmm2
        rsqrtss %xmm3, %xmm2
        movaps  %xmm2, %xmm4
        mulss   %xmm4, %xmm4
        mulss   %xmm3, %xmm4
        addss   .LCPI3_0(%rip), %xmm4
        mulss   .LCPI3_1(%rip), %xmm2
        mulss   x(%rip), %xmm0
        mulss   y(%rip), %xmm1
        addss   %xmm0, %xmm1
        mulss   %xmm2, %xmm1
        mulss   %xmm4, %xmm1
        ucomiss dcos(%rip), %xmm1
        seta    %al

VinInn replied:
btw that was clang
gcc (even 6.0) produces
```
inside(float, float):
        movaps  %xmm0, %xmm2
        movaps  %xmm1, %xmm3
        mulss   %xmm0, %xmm2
        mulss   %xmm1, %xmm3
        mulss   x(%rip), %xmm0
        mulss   y(%rip), %xmm1
        addss   %xmm3, %xmm2
        sqrtss  %xmm2, %xmm2
        mulss   dcos(%rip), %xmm2
        addss   %xmm1, %xmm0
        comiss  %xmm2, %xmm0
        seta    %al
        ret
```
one needs to add ```-mrecip`` to get a code similar to those from clang
There is a long argument if scalar sqrt is really slower than rsqrt +NR
again much depends of the processor ability to pipeline and parallelize fp instructions...
given that we are even questioning if Ofast is worth, the whole discussion is a bit lame

slava77 replied:
On my side, the discussion was actually triggered by the case of division by 0.
The rhs dcos*sqrt doesn't have x/0

VinInn replied:
so now we have 0<0 instead of Nan<0
always false even with Ofast...

slava77 replied:
0>0 is not an FPE.

VinInn commented:
@cmsbuild , please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 5d7fbc1c2a85014a1a83dde4a5b3787fbd8c4cc9
I found errors in the following unit tests: 
 
 ---> test DataFormatsIntervals_t had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13914/12312/summary.html 
  'List

VinInn committed:
avoid problems with rounding close to zero (flip sign)

cmsbuild commented:
Pull request #13914 was updated. @civanch, @cvuosalo, @mdhildreth, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

VinInn commented:
@cmsbuild, please test
the problem was with the test, the two algo may disagree fi the tested value is (inside tolerances) close to one extreme (was 0 and -0.1e-6)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13914/12312/summary.html

cmsbuild commented:
-1 
Tested at: f8cf2ad882ad9192640f7729b8b251b5d8e179e1
 When I ran the RelVals I found an error in the following worklfows: 
 135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13914/12326/summary.html 
  'List

VinInn commented:
interesting
```
Thread 2 (Thread 0x7fa3dcfbf700 (LWP 5572)):
#0  0x000000313bc0f37d in waitpid () from /lib64/libpthread.so.0
#1  0x00007fa3eba1a537 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/pluginFWCoreServicesPlugins.so
#2  0x00007fa3eba1ade2 in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/pluginFWCoreServicesPlugins.so
#3  0x00007fa3ee614a50 in std::(anonymous namespace)::execute_native_thread_routine (__p=<optimized out>) at ../../../../../libstdc++-v3/src/c++11/thread.cc:84
#4  0x000000313bc07aa1 in start_thread () from /lib64/libpthread.so.0
#5  0x000000313b8e893d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7fa3ee2b0ba0 (LWP 5339)):
#0  0x000000313b8df113 in poll () from /lib64/libc.so.6
#1  0x00007fa3eba1ac44 in full_read.constprop () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/pluginFWCoreServicesPlugins.so
#2  0x00007fa3eba1aeaa in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/pluginFWCoreServicesPlugins.so
#3  0x00007fa3eba1b02b in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00007fa3d47d7533 in DetectPatterns(ZonesOutput) () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/libL1TriggerL1TMuonEndCap.so
#6  0x00007fa3d47d809d in Patterns(std::vector<ZonesOutput, std::allocator<ZonesOutput> >) () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/libL1TriggerL1TMuonEndCap.so
#7  0x00007fa39818145f in L1TMuonEndCapTrackProducer::produce(edm::Event&, edm::EventSetup const&) () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/pluginL1TriggerL1TMuonEndCapPlugins.so
#8  0x00007fa3efb79771 in edm::EDProducer::doEvent(edm::EventPrincipal const&, edm::EventSetup const&, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/2016-16/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-04-10-0000/lib/slc6_amd64_gcc530/libFWCoreFramework.so
#9  0x00007fa3efc1a52f in edm::WorkerT<edm::EDProducer>::implDo(edm::EventPrincipal const&, edm::EventSetup const&, edm::ModuleCa
```
hard to believe that it is due to this PR...

davidlt commented:
It's not, it's broken in IBs.

davidlange6 commented:
actually this workflow is not broken in the IBs.. but indeed this error comes up in one workflow, so its presumably related -the L1 team is investigating.

> On Apr 13, 2016, at 9:25 AM, davidlt <notifications@github.com> wrote:
> 
> It's not, it's broken in IBs.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

cvuosalo commented:
+1

For #13914 f8cf2ad882ad9192640f7729b8b251b5d8e179e1

Bug fix in checking phi ranges for pixel hits.

The code changes are satisfactory. Jenkins tests against baseline CMSSW_8_1_X_2016-04-11-2300 show numerous tiny differences, but they are not significant. The Jenkins test error is not related to this PR.

Extended tests of workflow 25202.0_TTbar_13 with 70 events and workflows 1317.0_SingleElectronPt35_UP15 and 1318.0_SingleGammaPt10_UP15 with 1000 events each against baseline CMSSW_8_1_X_2016-04-03-1100 also show numerous tiny differences, but again they are not significant.

VinInn commented:
Simulation signed a previous version. the relval failure is not related to this PR
